### DISCUSSION TITLE
[lit/ssr]: `slotchange` event does not fire if the element has been SSRed.

### DISCUSSION THREAD

**Discussion by KonnorRogers** [lit/lit#4697] [Q&A]
When I use Lit SSR and have a `slotchange` events will not fire on first render.

I'm not sure if this belongs as a bug or discussion, as I understand the browser does not do this either.

https://codepen.io/paramagicdev/pen/NWZPGXO

Some ideas here:

- on `firstUpdated`, query all shadow root slot elements and fire a `slotchange` event. This may not be 100% representative of the browser if theres no content on SSR and on client hydration, but slotchange handlers _should_ probably idempotent anyways.

- detect in templates any `@slotchange` events, but this will miss any imperative listeners set in a constructor for example.

- Leave it to a user to handle. Right now I have a really hacky solution:

```js
export default class ShoelaceElement extends LitElement {
  constructor () {
    super()
    // if `this.shadowRoot` exists in the constructor, we know this element was SSRed.
    if (this.shadowRoot) {
      this.didSSR = true
    }
  }
  
  firstUpdated () {
    if (this.didSSR) {
      this.shadowRoot.querySelectorAll("slot").forEach((el) => {
        // "slotchange" bubbles, but does not cross shadow roots.
        el.dispatchEvent(new Event("slotchange", { bubbles: true, composed: false, cancelable: false ))
      })
    }
  }
}
```

Am I missing something obvious here? Is there another solution I hadn't thought of? Is there a better way to detect if the element has been SSRed by Lit?

**Comment by sorvell** [comment]
Yeah, this would occur for any lazily added `slotchange` event listener. It's not unique to SSR, but practically speaking that's probably the only time you'd encounter it.

Here is a quick idea of how to handle the situation somewhat generally:
1. Create a directive that takes the event listener.
2. Have the directive try to detect if the event should be synthetically dispatched. I think this can be done by checking (1) if the host's `hasUpdated` is false, and (2) if the part element `isConnected` is true.
3. Synthetically dispatch the event if so. And this should probably be async just to make the part handling simpler.

Here's a [quick example](https://lit.dev/playground/#gist=fbd98c7e19c79294df2a27b649afee12).

Does this cover the use case?

**Reply by KonnorRogers** [reply]
🤔I don't particularly love having to rewrite every slotchange handler to have a directive, but it does seem to fit the use case.

**Reply by justinfagnani** [reply]
Maybe we can do something in the hydrate script
